### ISSUE TITLE
Devices still show in cacti despite being deleted 

### ISSUE THREAD

**Issue by bmfmancini** [Cacti/cacti#2937]
Hey Guys 

So I deleted about 100 Devices to re-inport them via automation into cacti 
when I  try to re-import cacti skips the IP's reporting device already in cacti 

I tried the remove_device script  and used a regex which found all the sites I used --confirm the script reports back successful 

yet AutoM8 still skips the IP's and if you run the script again the devices are still there !!

**Comment by bmfmancini** [comment]
its weird I did a full sync as well and that didn't seem to help 
I restored from a DB backup to resolve for now but its something for sure to look into

**Comment by cigamit** [comment]
The only time that this could possibly happens is when: 

1) When devices are removed while the remote poller is running
2) When the main data collector can not reach the remote database over a mysql connection.

Case 1) is difficult to completely overcome.  We do have a purge command that happens at the bottom of each poller run, but if the remote database is not accessible, the purge will not complete successfully.

**Comment by bmfmancini** [comment]
It's got me scratching my head because even the remove device php script
comes back as successful

I ran the script on both the main and remote pollers as well

The main poller found all of the devices that I was looking for I ran with
a --confirm result came back as successful

The devices did not show up in the device group and red cleaner marked the
rrrds as orphaned

**Comment by cigamit** [comment]
From the main system, edit the Remote Data Collector and test the connection.  Is everything coming back successful?

**Comment by bmfmancini** [comment]
I'll check that tomorrow and let you know

Thanks!

**Comment by bmfmancini** [comment]
Just a thought though

If the remote pollers couldn't access the main poller wouldn't I have
bigger issues
The graphs from that poller work just fine and that poller gets all the
settings properly even when updated at the main poller

Also there are no errors with poller syncs

**Comment by bmfmancini** [comment]
Ok tested 

The main poller is able to reach the remote pollers DB just fine 
tested by messing around with the credentials 

Still very weird

**Comment by bmfmancini** [comment]
Ok so I revisited this and something is still no right 

When you delete the device off the main poller it is gone you can add the site back manually but through AUTOM8 is will show as still in cacti 

If you select the hostname table the entry is still in the DB ! 

The main poller can reach the remote pollers just fine replication works as well 
no sql errors in the log and no errors in the mariadb log either

**Comment by netniV** [comment]
That is likely that AUTOM8 isn't ignoring devices that are set to deleted in the hosts table.  Because of replication, they are not totally removed until all replication is complete.

**Comment by bmfmancini** [comment]
I forced replication and still the same
It's like the table is not actually replicating

**Comment by bmfmancini** [comment]
Deleted a device directly from the DB 

here is the command I ran 

(delete  from  host where hostname like 'Ip address;)

I re-ran autom8 and it picked up the device and re-added it 
I did notice duplicate records of the device that I added manually I don't think cacti is marking the device as deleted 

No errors are seen from the SQL call and no troubles issuing the command as the same user that cacti uses so not a permissions issue 

I am not sure if this is specific to 1.2.5

**Comment by netniV** [comment]
Deleting the device manually will leave behind datasources and graphs related to it.  It will also not sync to the other pollers.  You should make sure that you don't have rows in other tables (especially tables with a host_id) that match.  If you do, you'll have to start cascading through the tables or you could end up with other issues were code starts at a data source or graph and works backwards.

**Comment by netniV** [comment]
I would do a select on the device you want to delete manually before you do to keep a record of it.

**Comment by bmfmancini** [comment]
Thanks

I'm not planning on doing this regularly because cacti should be deleting
devices properly

I just wanted to be sure all permission were fine

Still need to figure out why cacti is not doing this properly

**Comment by bmfmancini** [comment]
This is still ongoing I have run the audit db script all good analyze ab all good 

I still see entries in the db from deleted devices  I don't think the entries ate being marked for deletion

**Comment by bmfmancini** [comment]
Oh and this is on 1.2.7

**Comment by cigamit** [comment]
You need to be on branch 1.2.x.  Also, perform a diff between config.php.dist and config.php, let me know what is different.  I really suspect that you are having some database communication problems.  Are you sure you don't have a database issue?  Test connection comes back good for all collectors?

**Comment by bmfmancini** [comment]
I will do the diff and get back to you

Everything works I can add devices even on remote pollers no issue

Clicking the test button works fine

Mysql from cli work both from remote to.main and main to remote

Poller replication works no errors reported

I would think if it was a db connect issue it would have manifested in
other ways though

**Comment by cigamit** [comment]
How large is the install?  What's you lowest collection frequency?  I'll do some testing now that I've installed a remote data collector.

**Comment by bmfmancini** [comment]
Ok so the install we have has 2100 devices

With 1 main and 2 remote pollers

1 minute polling interveral

I don't have the polling stats right now

**Comment by bmfmancini** [comment]
Ok a fresh example 

multi pollers setup 1.2.7  not the same as the above examples 
I deleted 2 devices cacti shows 0 devices to poll yet DB shows the 2 devices and templates show 2 devices using them see below


![image](https://user-images.githubusercontent.com/13388748/68724754-5f41e080-058a-11ea-9e70-05aa1556671c.png)


![image](https://user-images.githubusercontent.com/13388748/68724780-71238380-058a-11ea-93c8-f08bb3dd8341.png)


![image](https://user-images.githubusercontent.com/13388748/68724812-8d272500-058a-11ea-9c56-e0e3c858343a.png)

I forced a sync between the pollers here is the output


2019/11/12 20:19:45 - WEBUI NOTE: All selected Remote Data Collectors in [2] synchronized correctly by user admin
--
2019/11/12 20:19:45 - CMDPHP NOTE: Table data_input_data Not Replicated to Remote Poller 2 Due to No Rows Found
2019/11/12 20:19:45 - CMDPHP NOTE: Table graph_templates_item Not Replicated to Remote Poller 2 Due to No Rows Found
2019/11/12 20:19:44 - CMDPHP NOTE: Table data_template_rrd Not Replicated to Remote Poller 2 Due to No Rows Found
2019/11/12 20:19:44 - CMDPHP NOTE: Table data_template_data Not Replicated to Remote Poller 2 Due to No Rows Found
2019/11/12 20:19:44 - CMDPHP NOTE: Table graph_local Replicated to Remote Poller 2 With 31 Rows Updated
2019/11/12 20:19:44 - CMDPHP NOTE: Table data_local Not Replicated to Remote Poller 2 Due to No Rows Found
2019/11/12 20:19:44 - CMDPHP NOTE: Table poller_reindex Not Replicated to Remote Poller 2 Due to No Rows Found
2019/11/12 20:19:44 - CMDPHP NOTE: Table host_snmp_cache Not Replicated to Remote Poller 2 Due to No Rows Found
2019/11/12 20:19:44 - CMDPHP NOTE: Table host Replicated to Remote Poller 2 With 2 Rows Updated
2019/11/12 20:19:43 - CMDPHP NOTE: Table poller_item Not Replicated to Remote Poller 2 Due to No Rows Found
2019/11/12 20:19:43 - CMDPHP NOTE: Table poller_command Not Replicated to Remote Poller 2 Due to No Rows Found
2019/11/12 20:19:43 - CMDPHP NOTE: Table host_snmp_query Not Replicated to Remote Poller 2 Due to No Rows Found
2019/11/12 20:19:43 - CMDPHP NOTE: Table user_domains_ldap Not Replicated to Remote Poller 2 Due to No Rows Found
2019/11/12 20:19:43 - CMDPHP NOTE: Table user_domains Not Replicated to Remote Poller 2 Due to No Rows Found
2019/11/12 20:19:43 - CMDPHP NOTE: Table user_auth_realm Replicated to Remote Poller 2 With 28 Rows Updated
2019/11/12 20:19:43 - CMDPHP NOTE: Table user_auth_group_realm Not Replicated to Remote Poller 2 Due to No Rows Found
2019/11/12 20:19:43 - CMDPHP NOTE: Table user_auth_group_perms Not Replicated to Remote Poller 2 Due to No Rows Found
2019/11/12 20:19:42 - CMDPHP NOTE: Table user_auth_group_members Not Replicated to Remote Poller 2 Due to No Rows Found
2019/11/12 20:19:42 - CMDPHP NOTE: Table user_auth_group Not Replicated to Remote Poller 2 Due to No Rows Found
2019/11/12 20:19:42 - CMDPHP NOTE: Table user_auth Replicated to Remote Poller 2 With 2 Rows Updated
2019/11/12 20:19:42 - CMDPHP NOTE: Table version Replicated to Remote Poller 2 With 1 Rows Updated
2019/11/12 20:19:42 - CMDPHP NOTE: Table poller Replicated to Remote Poller 2 With 2 Rows Updated
2019/11/12 20:19:42 - CMDPHP NOTE: Table data_input_fields Replicated to Remote Poller 2 With 98 Rows Updated
2019/11/12 20:19:42 - CMDPHP NOTE: Table snmp_query Replicated to Remote Poller 2 With 9 Rows Updated
2019/11/12 20:19:42 - CMDPHP NOTE: Table host_template_snmp_query Replicated to Remote Poller 2 With 14 Rows Updated
2019/11/12 20:19:42 - CMDPHP NOTE: Table host_template_graph Replicated to Remote Poller 2 With 31 Rows Updated
2019/11/12 20:19:41 - CMDPHP NOTE: Table host_template Replicated to Remote Poller 2 With 6 Rows Updated
2019/11/12 20:19:41 - CMDPHP NOTE: Table data_input Replicated to Remote Poller 2 With 31 Rows Updated

the remote poller shows 2 devices assigned to it ???

![image](https://user-images.githubusercontent.com/13388748/68724962-0292f580-058b-11ea-813e-ccf25907cdb0.png)

Second poller device view

![image](https://user-images.githubusercontent.com/13388748/68725058-4980eb00-058b-11ea-8314-d4c0d1fad38f.png)

See below for sql test

![bi-directional](https://user-images.githubusercontent.com/13388748/68725183-94026780-058b-11ea-809b-3709c2910e71.gif)


[root@localhost ~]# mysql -u cacti -p -h 192.168.1.228
Enter password: 
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 15916
Server version: 5.5.64-MariaDB MariaDB Server

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> use cacti;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
MariaDB [cacti]> quit
Bye
[root@localhost ~]# ^C



MariaDB [cacti]> select hostname  from host;
+---------------+
| hostname      |
+---------------+
| 192.168.1.253 |
| 4.2.2.2       |
+---------------+
2 rows in set (0.00 sec)

**Comment by bmfmancini** [comment]
Hey Guys 

I am still seeing this  any new ideas ?
I think I have provided all that I can even the LAB is doing it 
I belive the tables are not replicating to reflect that the device is deleted to the remote pollers as this doesnt happen with the main poller

**Comment by bmfmancini** [comment]
Hey guys

Wondering if you have had a chance to look at this  any info I can provide please let me know !

**Comment by bmfmancini** [comment]
so funny thing I found  when I delete the device it does not show up in the device list but shows up everywhere else 

It shows up in the auto-complete dropdown when you search for it in the device view but if you click it nothing shows up 

In the monitor plugin tab the device shows in the device view 
If you try to re-add the device via autom8 the log shows the device as already in cacti
if you add the device manually the device shows up twice in the auto complete dropdown 

please I am at a loss here any help would be appreciated

**Comment by bmfmancini** [comment]
Oh btw on the recent device deletion I forced a poller sync as well same outcome

**Comment by netniV** [comment]
This is likely because we introduced the 'deleted' column to aid with remote poller syncs.  After all, if an entry on the main poller isn't there, the remotes will re-add it since it's considered a new device.

Monitor needs to ignore them, as does practically everything else that references the hosts table.

**Comment by bmfmancini** [comment]
so that would also explain why autom8 sees the device as already in cacti ?
so is this a bug in the logic ?

**Comment by netniV** [comment]
Well, that is a catch 22... if we have a device that marked as being deleted (deleted = 'on') then should automate be adding it? The answer is probably not.  However, how long after being deleted, should the device remain in the host table.  For that, I'm not sure how long it does last for without reviewing the code again but I suspect when the poller the device belongs to, picks up that the device is deleted and removes it's local version, then the remote one, it'll go.

**Comment by bmfmancini** [comment]
the example I have is this 

We have a device in cacti that was decomissioned so removed the device 
a day later we re-used the IP so attempted to add in via automation thats when I noticed the problem with autom8 

maybe have an option to allow autom8 re-add a deleted device during a network scan ?

**Comment by bmfmancini** [comment]
Ok so indeed when I do a select for items marked deleted I  see the entries !!!
is there any way I can manually delete these to solve my issue for now ?

I mean a clean way via a sql  since you cant do it via the gui  
BTW one of these devices has been in the DB marked as delete for about a month  so I think the routine thats supposed to clear this out is not working

**Comment by netniV** [comment]
Now that we need to do some proper testing against unless @cigamit or @TheWitness knows of a way that cleans that automatically?  Maybe that just isn't kicking in for you for some reason?

**Comment by bmfmancini** [comment]
Yeah I am not sure I know that in 1.2.4 this was not an issue not quite sure when the deleted column was introduced when I was doing testing for autom8 we would delete and re-add devices a lot without issues 

I am sure that something in the background is not picking up the tag I belive its also happening on the main poller as well so I dont think its a replication issue as well but I will test again just to be sure

**Comment by bmfmancini** [comment]
I have a theory could this be THOLD holding on to the record?
The reason I say is that I have a device that was deleted 50 days ago that still shows as a down device in thold now if you click edit device all of the info is present in the device page IP and description but no graphs or data sources 

The device table in the DB does indeed have this set as a deleted device but why even after 50 days does it insist on staying ??

**Comment by bmfmancini** [comment]
Hey Guys 

While we look into this would you be able to tell me of a manual way to clear these entries ?

**Comment by bmfmancini** [comment]
I found this block of code in api_device

function api_device_purge_deleted_devices() {
        $devices = db_fetch_assoc_prepared('SELECT id, poller_id
                FROM host
                WHERE deleted = "on"
                AND UNIX_TIMESTAMP(last_updated) < UNIX_TIMESTAMP()-86400');

So this would mean that after 24 hours that the devices should be purged I belive something is wrong with this block since I have seen devices both in my production and multiple labs stick around for days

**Comment by TheWitness** [comment]
Run this, post the output:

```sql
SELECT id, deleted, last_updated FROM host WHERE deleted="on";
```

In general, I don't know why there is a library include inside of lib/api_device.php.  That is simply not right, especially if it's not within a function definition.

**Comment by TheWitness** [comment]
I've removed the include in lib/api_device.php.  But to apply it, you have to take all the files.  It should be pretty harmless.

**Comment by bmfmancini** [comment]
Hey thanks!
I will try it do you mean to take all the files in lib?

**Comment by bmfmancini** [comment]
Here is the output from the db

SELECT id, deleted, last_updated FROM host WHERE deleted="on";
+------+---------+---------------------+
| id   | deleted | last_updated        |
+------+---------+---------------------+
|  996 | on      | 2020-01-22 09:41:11 |
| 1405 | on      | 2020-01-22 09:41:06 |
| 1430 | on      | 2020-01-22 09:41:06 |
| 1444 | on      | 2020-01-22 09:41:07 |
| 2404 | on      | 2020-01-22 09:41:10 |
| 2416 | on      | 2020-01-22 09:41:06 |
| 2428 | on      | 2020-01-22 09:41:06 |
| 2437 | on      | 2020-01-22 09:41:05 |
| 2442 | on      | 2020-01-22 09:41:05 |
| 2445 | on      | 2020-01-22 09:41:08 |
| 2458 | on      | 2020-01-22 09:41:11 |
| 2461 | on      | 2020-01-22 09:41:05 |
| 2462 | on      | 2020-01-22 09:41:09 |
| 2472 | on      | 2020-01-22 09:41:08 |
| 2478 | on      | 2020-01-22 09:41:09 |
| 2553 | on      | 2020-01-22 09:41:05 |
| 2747 | on      | 2020-01-22 09:41:04 |
| 2750 | on      | 2020-01-22 09:41:03 |
| 2756 | on      | 2020-01-22 09:41:09 |
| 2937 | on      | 2020-01-22 09:41:06 |
| 2938 | on      | 2020-01-22 09:41:09 |
| 2939 | on      | 2020-01-22 09:41:12 |
| 2940 | on      | 2020-01-22 09:41:11 |
| 2941 | on      | 2020-01-22 09:41:05 |
| 2942 | on      | 2020-01-22 09:41:13 |
| 2943 | on      | 2020-01-22 09:41:08 |
| 2944 | on      | 2020-01-22 09:41:11 |
| 2945 | on      | 2020-01-22 09:41:13 |
| 2946 | on      | 2020-01-22 09:41:24 |
| 2947 | on      | 2020-01-22 09:41:05 |
| 2948 | on      | 2020-01-22 09:41:13 |
| 2949 | on      | 2020-01-22 09:41:04 |
| 2995 | on      | 2020-01-22 09:40:34 |
| 2998 | on      | 2020-01-22 09:41:17 |
| 3050 | on      | 2020-01-22 09:41:10 |
| 3100 | on      | 2020-01-22 09:41:06 |
| 3101 | on      | 2020-01-22 09:41:05 |
| 3184 | on      | 2020-01-22 09:41:09 |
| 3222 | on      | 2020-01-22 09:41:10 |
| 3269 | on      | 2020-01-22 09:41:07 |
| 3284 | on      | 2020-01-22 09:41:04 |
| 3300 | on      | 2020-01-22 09:41:08 |
| 3777 | on      | 2020-01-22 09:41:06 |
| 4498 | on      | 2020-01-22 09:41:06 |
| 4533 | on      | 2020-01-22 09:41:08 |
| 4564 | on      | 2020-01-22 09:41:08 |
| 4609 | on      | 2020-01-22 09:41:09 |
| 4640 | on      | 2020-01-22 09:41:06 |
| 4728 | on      | 2020-01-22 09:41:06 |
| 4756 | on      | 2020-01-22 09:41:05 |
| 4813 | on      | 2020-01-22 09:41:05 |
+------+---------+---------------------+

**Comment by TheWitness** [comment]
Well, that explain it then.  Just need to find out where the update is occurring.

**Comment by bmfmancini** [comment]
I applied the new files autom8 is working fine no errors seen as of yet

**Comment by bmfmancini** [comment]
@TheWitness  the update is happening at each poll
seems to update each minute

**Comment by bmfmancini** [comment]
Ok so I added a device via autom8 now and deleted it its been 5 mins and the last updated shows as of 5 mins ago so something is definitely different the odd thing is that the other devices still show updating currently

**Comment by bmfmancini** [comment]
| 4756 | on      | 2020-01-22 10:36:05 |
| 4813 | on      | 2020-01-22 10:36:05 |
| 4908 | on      | 2020-01-22 10:26:17 |
+------+---------+---------------------+
52 rows in set (0.00 sec)

**Comment by TheWitness** [comment]
I'm in the process of reviewing code that makes updates to the `host` table.  There will be an update later this evening to cover some opportunities where updates may come in, say for example, as mentioned, with autom8.

**Comment by TheWitness** [comment]
Okay, think this is fixed now.  There is a spine update too.  Should be committed within a few minutes.

**Comment by bmfmancini** [comment]
Dude thanks so much I will definitely test this and get back to you !

**Comment by cigamit** [comment]
So, I did some testing and found that discovery that was made on a remote data collector did not get pushed to the local databases, forcing a full sync to resolve that.  The commit that I just made ensures that you don't have to full sync to make the discovery work.  This this things goose is cooked now.  I also tightened the screws on the purge timing to just under 10 minutes.

**Comment by ddb4github** [comment]
Previous commit rollback https://github.com/Cacti/cacti/issues/2126, and impact all caller of API api_device_save and push_out_host. The caller have to explicitly include lib/poller.php

**Comment by netniV** [comment]
Not sure where you are going with that?

**Comment by bmfmancini** [comment]
On the original issue, I am seeing that this seems to be fixed in 1.2.9 in my lab testing anyways
